pymc_marketing/mmm/models/components/lagging.py

```diff
@@ -0,0 +1,394 @@
+from typing import Dict, Optional, Union
+
+import pymc as pm
+
+from pymc_marketing.mmm.transformers import geometric_adstock, weibull_adstock
+from pymc_marketing.mmm.utils import _get_distribution
+
+
+class BaseFunction:
+    def __init__(self, max_lagging: int = 4, model: Optional[pm.Model] = None):
+        self.max_lagging = max_lagging
+        self.model = model
+
+    def apply(self, data):
+        raise NotImplementedError("Subclasses must implement this method.")
+
+
+class GeometricAdstockComponent(BaseFunction):
+    """
+    Geometric Adstock Component.
+
+    This component applies geometric adstock transformation to the input data.
+    It calculates the adstock effect of past marketing efforts on the current response.
+
+    Parameters:
+    -----------
+    max_lagging : int
+        The maximum lagging period to consider for adstock effect.
+    model : pymc.Model
+        The PyMC model object.
+    model_config : dict
+        The configuration dictionary for the model.
+
+    Attributes:
+    -----------
+    REQUIRED_KEYS : list
+        The list of required keys in the model_config dictionary.
+
+    Methods:
+    --------
+    apply(data: Union[pm.Data, pm.MutableData]) -> pm.Deterministic:
+        Apply the geometric adstock transformation to the input data.
+
+    """
+
+    REQUIRED_KEYS = [
+        "adstock_alpha",
+    ]
+
+    def __init__(
+        self,
+        max_lagging: int = 4,
+        model: Optional[pm.Model] = None,
+        model_config: Optional[Dict] = None,
+    ):
+        self.max_lagging = max_lagging
+        self.model = model
+        if model_config is not None and not all(
+            key in model_config for key in self.REQUIRED_KEYS
+        ):
+            self.model_config = {**model_config, **self._default_lagging_config}
+        elif model_config is None:
+            self.model_config = self._default_lagging_config
+        else:
+            self.model_config = model_config
```

Comment:
Seems to be logic that you apply in every class. Can likely put it in the baseclass